### PR TITLE
Rust add snmp 02

### PR DESCRIPTION
**Pull request by chifflier**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2738

This is an update of #3581

Describe changes (since last PR):
- Upgrade to nom4 (inner parser crate was also updated)
- Rebase to match changes
- Rename keywords (snmp_version -> snmp.version, etc.)

There will also be a matching suricata-verify pull request with test files.

### CODE REVIEW

File: doc/userguide/rules/snmp-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
+SNMP keywords
+=============
+
+snmp_version
```
Comment (victorjulien) [inline doc/userguide/rules/snmp-keywords.rst]: syntax is now snmp.version

File: rust/src/snmp/detect.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/* Copyright (C) 2017-2018 Open Information Security Foundation
```
Comment (victorjulien) [inline rust/src/snmp/detect.rs]: 2019

File: rust/src/snmp/log.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
+/* Copyright (C) 2018 Open Information Security Foundation
```
Comment (victorjulien) [inline rust/src/snmp/log.rs]: 2019

File: src/app-layer-snmp.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
+/* Copyright (C) 2015 Open Information Security Foundation
```
Comment (victorjulien) [inline src/app-layer-snmp.c]: 2019 (can you check the other files as well?)

File: src/app-layer-snmp.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
+ */
+
+#ifndef __APP_LAYER_SNMP_H__
+#define __APP_LAYER_SNMP_H__
+
+void RegisterSNMPParsers(void);
+void SNMPParserRegisterTests(void);
```
Comment (victorjulien) [inline src/app-layer-snmp.h]: No need to export this I think

File: src/detect-snmp-community.c [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].Setup =
+        DetectSNMPCommunitySetup;
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].RegisterTests =
+        DetectSNMPCommunityRegisterTests;
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].url = DOC_URL DOC_VERSION "/rules/snmp-keywords.html#snmp_community";
+
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].flags |= SIGMATCH_NOOPT|SIGMATCH_INFO_STICKY_BUFFER;
+
+    /* register inspect engines */
+    DetectAppLayerInspectEngineRegister("snmp_community",
```
Comment (victorjulien) [inline src/detect-snmp-community.c]: please use the v2 version of this

File: src/detect-snmp-community.c [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+    const Signature *s, const SigMatchData *smd,
+    Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id);
+static void DetectSNMPCommunityRegisterTests(void);
+static int g_snmp_rust_id = 0;
+
+void DetectSNMPCommunityRegister(void)
+{
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].name = "snmp.community";
+    sigmatch_table[DETECT_AL_SNMP_COMMUNITY].alias = "snmp_community";
```
Comment (victorjulien) [inline src/detect-snmp-community.c]: this alias can be removed. It's used to provide backward compat in other cases, but here it's not need as the code and functionality is all new

File: src/detect-snmp-community.c (hunk 6 above)
Comment (victorjulien) [reply src/detect-snmp-community.c]: plus adding mpm support would be great

File: src/detect-snmp-community.c [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    if (data != NULL) {
+        ret = DetectEngineContentInspection(de_ctx, det_ctx, s, smd,
+            NULL, f, (uint8_t *)data, data_len, 0, DETECT_CI_FLAGS_SINGLE,
+            DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE);
+    }
+
+    return ret;
+}
+
+#ifdef UNITTESTS
```
Comment (victorjulien) [inline src/detect-snmp-community.c]: can you move the tests to src/tests/detect-snmp-community.c? E.g. see #3853

File: src/detect-snmp-community.c [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+    /* Check that we have app-layer state. */
+    FAIL_IF_NULL(f.alstate);
+
+    SigMatchSignatures(&tv, de_ctx, det_ctx, p);
+    FAIL_IF(!PacketAlertCheck(p, 1));
+    FAIL_IF(PacketAlertCheck(p, 2));
+
+    /* Cleanup. */
+    if (alp_tctx != NULL)
```
Comment (victorjulien) [inline src/detect-snmp-community.c]: all these conditions should be unnecessary if the 'FAIL_IF' logic above is correct

File: src/detect-snmp-community.c [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        "alert snmp any any -> any any ("
+        "msg:\"SNMP Test Rule\"; "
+        "snmp_community; content:\"private\"; "
+        "sid:2; rev:1;)");
+    FAIL_IF_NULL(s);
+
+    SigGroupBuild(de_ctx);
+    DetectEngineThreadCtxInit(&tv, (void *)de_ctx, (void *)&det_ctx);
+
+    FLOWLOCK_WRLOCK(&f);
```
Comment (victorjulien) [inline src/detect-snmp-community.c]: no need to use this in tests

File: src/detect-snmp-pdu_type.c [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+static int DetectSNMPPduTypeMatch (ThreadVars *, DetectEngineThreadCtx *, Flow *,
+                                   uint8_t, void *, void *, const Signature *,
+                                   const SigMatchCtx *);
+
+void DetectSNMPPduTypeRegister(void)
+{
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].name = "snmp_pdu_type";
```
Comment (victorjulien) [inline src/detect-snmp-pdu_type.c]: only set name with the dot notation

File: src/detect-snmp-pdu_type.c [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].alias = "snmp.pdu_type";
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].desc = "match SNMP Pdu type";
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].url = DOC_URL DOC_VERSION "/rules/snmp-keywords.html#snmp_pdu_type";
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].Match = NULL;
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].AppLayerTxMatch = DetectSNMPPduTypeMatch;
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].Setup = DetectSNMPPduTypeSetup;
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].Free = DetectSNMPPduTypeFree;
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].RegisterTests = DetectSNMPPduTypeRegisterTests;
+    sigmatch_table[DETECT_AL_SNMP_PDU_TYPE].flags |= SIGMATCH_NOOPT|SIGMATCH_INFO_STICKY_BUFFER;
```
Comment (victorjulien) [inline src/detect-snmp-pdu_type.c]: this is not considered a sticky buffer

File: src/detect-snmp-pdu_type.c [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+ *
+ * \param de_ptr Pointer to DetectSNMPPduTypeData.
+ */
+static void DetectSNMPPduTypeFree(void *ptr)
+{
+    SCFree(ptr);
+}
+
+
+#ifdef UNITTESTS
```
Comment (victorjulien) [inline src/detect-snmp-pdu_type.c]: move to own file as well

File: src/detect-snmp-version.c [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+                                   uint8_t, void *, void *, const Signature *,
+                                   const SigMatchCtx *);
+
+/**
+ * \brief Registration function for snmp_procedure keyword.
+ */
+void DetectSNMPVersionRegister (void)
+{
+    sigmatch_table[DETECT_AL_SNMP_VERSION].name = "snmp.version";
+    sigmatch_table[DETECT_AL_SNMP_VERSION].alias = "snmp_version";
```
Comment (victorjulien) [inline src/detect-snmp-version.c]: no need for this alias

File: src/detect-snmp-version.c [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+    sigmatch_table[DETECT_AL_SNMP_VERSION].desc = "match SNMP version";
+    sigmatch_table[DETECT_AL_SNMP_VERSION].url = DOC_URL DOC_VERSION "/rules/snmp-keywords.html#snmp_version";
+    sigmatch_table[DETECT_AL_SNMP_VERSION].Match = NULL;
+    sigmatch_table[DETECT_AL_SNMP_VERSION].AppLayerTxMatch = DetectSNMPVersionMatch;
+    sigmatch_table[DETECT_AL_SNMP_VERSION].Setup = DetectSNMPVersionSetup;
+    sigmatch_table[DETECT_AL_SNMP_VERSION].Free = DetectSNMPVersionFree;
+    sigmatch_table[DETECT_AL_SNMP_VERSION].RegisterTests = DetectSNMPVersionRegisterTests;
+    sigmatch_table[DETECT_AL_SNMP_VERSION].flags |= SIGMATCH_NOOPT|SIGMATCH_INFO_STICKY_BUFFER;
```
Comment (victorjulien) [inline src/detect-snmp-version.c]: not a sb

File: src/detect-snmp-version.c [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+ *
+ * \param de_ptr Pointer to DetectSNMPVersionData.
+ */
+static void DetectSNMPVersionFree(void *ptr)
+{
+    SCFree(ptr);
+}
+
+
+#ifdef UNITTESTS
```
Comment (victorjulien) [inline src/detect-snmp-version.c]: please move these

**Comment by chifflier** [comment]
Thanks, all requested changes merged.
Closing pr, superseded by  #3883

**Pull request closed without merging**